### PR TITLE
STREAMS: Multiple topics extractor proposal

### PR DESCRIPTION
**Pull request by tdanylchuk**
Approach of adding possibility to extract several topics for flexible message routing.

I didn't find "nice" way how to apply such feature and keep all the changes backward compatible so I decided to ask if this change is acceptable via creating a PR. If it's OK - I'll finish all needed stuff (documentation and tests)

This would be awesome change, which will allow to chose several topics for single message dynamically based on key/value or application configuration.

Please let me know should I go further with this change and polish it.

### CODE REVIEW

**Comment by mjsax** [comment]
@tdanylchuk Thanks for the PR -- as mentioned on the Jira, this is a public API change and requires a KIP: https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals

**Comment by tdanylchuk** [comment]
Am I ask you that this change has a good chance to be applied ?) If KIP will be submitted, of course.

**Comment by mjsax** [comment]
I think it would be useful. However, I am just one vote. The only way to find out, is to propose the KIP.

We could also add this for the case without topic extractor maybe. Instead of
```
stream.to("topic1");
stream.to("topic2");
```
We could allow:
```
stream.to("topic1", "topic2");
```

**Comment by vvcephei** [comment]
Hi @tdanylchuk ,

Just to add to what @mjsax suggested, it would be pretty straightforward to convert `KStream.to(String topic)` to `KStream.to(String... topic)`, so that should probably be part of the proposal.

As far as changing the `TopicNameExtractor` goes, it's a public interface, and there's no clean way to change the return from a scalar String to a collection or array of Strings without breaking any existing source code. Most likely, you'd need to propose to add a new interface and corresponding new overloads of `to` to call them.

The fact that people might be using lambdas further complicates the situation, since having two methods with the same name both taking interfaces with conflicting functional signatures (the arguments are the same, but the return type is different) causes ambiguity that the compiler can't resolve. You can cast the argument to resolve it, but since the cast isn't required today, it would again cause source code to break.

It might be easiest to just make a new method, something like `toAll(TopicNamesExtractor)` and just sidestep the whole problem.

In other words, to answer your initial question, yes, FWIW, I think it's a worthwhile feature, but we really do need to have a good backward compatibility story. But like @mjsax, I'm just one vote. (And actually I'm not really a vote anyway, since I'm not a committer ;) )

Thanks! I look forward to your KIP and the discussion.
-John

**Comment by mjsax** [comment]
Closing this stale PR due to inactivity. We never received the required KIP for this change.
